runtime/reconcilers/metrics_view.go

```diff
@@ -119,11 +122,69 @@ func (r *MetricsViewReconciler) Reconcile(ctx context.Context, n *runtimev1.Reso
 		mv.State.Streaming = !hasInternalRef
 	}
 
+	// set cache controls
+	olap, release, err := r.C.AcquireOLAP(ctx, mv.Spec.Connector)
+	if err != nil {
+		return runtime.ReconcileResult{Err: fmt.Errorf("failed to get OLAP: %w", err)}
+	}
+	defer release()
+
+	validSpec := mv.State.ValidSpec
+	if validSpec != nil && validSpec.Cache != nil && validSpec.Cache.Enabled != nil && *validSpec.Cache.Enabled &&
+		mv.State.Streaming && validSpec.Cache.KeySql == "" && watermarkForMetricsView(validSpec) == "" {
+		return runtime.ReconcileResult{Err: fmt.Errorf("cache enabled for streaming metrics view %q but no `watermark` or `key_sql` provided", self.Meta.Name)}
+	}
+
 	// Update state. Even if the validation result is unchanged, we always update the state to ensure the state version is incremented.
 	err = r.C.UpdateState(ctx, self.Meta.Name, self)
 	if err != nil {
 		return runtime.ReconcileResult{Err: err}
 	}
+	// Override the cache control with our calculated cache control
+	if validSpec != nil {
+		validSpec.Cache = metricsViewCacheControl(validSpec, mv.State.Streaming, self.Meta.StateUpdatedOn, olap.Dialect())
+	}
 
 	return runtime.ReconcileResult{Err: validateErr}
 }
+
+func metricsViewCacheControl(spec *runtimev1.MetricsViewSpec, streaming bool, updatedOn *timestamppb.Timestamp, dialect drivers.Dialect) *runtimev1.MetricsViewSpec_Cache {
```

Comment:
I think it would be better to move the logic (which basically just applies defaults) in this function to the `metricsview_cache_key` resolver – to keep the cache key logic in a single place. (It also makes refactoring the defaults easier because they will only be applied at runtime, not persisted in the DB as here.)

Reply:
1. We still need to validate for `key_sql` i.e. whether we are able to resolve a valid key_sql for streaming metrics views. It will also be ideal to do that validation in reconciler and not during query time. Since we are anyways doing some validations it made sense to also apply defaults.
2. In reconciler I always mark cache enabled or disabled which the resolvers can use to check if caching should be enabled and get `metricsViewCacheKey`. Now resolvers will need to query every time and handle an error thrown by `metrics_cache_key` resolver.

>It also makes refactoring the defaults easier because they will only be applied at runtime, not persisted in the DB as here

Since defaults can only be changed with a code change so it should make sure the defaults are also updated on metrics view reconcile on restart.

Having said that I also thought about keeping this in resolver and was indifferent to both. 
I can move this logic to resolver as well.

Reply:
By the way if we don't apply defaults in reconciler than we can't resolve this todo since executor can't call `metrics_cache_key` resolver. 
https://github.com/rilldata/rill/blob/ba6f2ba9d62b591db91f8bc1af202f64e68fa64e/runtime/metricsview/executor.go#L68-L72

Reply:
I can think of following ideas:
1. We discussed on resolving this via a refresh schedule on metrics_view. This could be evaluated. In that case we can compute the key and store it directly in spec.
2. Another option is to let runtime expose APIs to set/get data in query_cache which the executor can utilise to set/get `metrics_cache_key` in the cache. This could be useful for other use cases as well. Like as of now we are always querying watermark so it could be cached as well(May be it is not queried always, just thinking out aloud).

Reply:
Option 2 sounds most convenient to me. Good chance we may come up with similar use cases in the future as well.

Reply:
Right now we basically only have caching happening in `runtime.Runtime.Resolve` (and the legacy `runtime.Runtime.Query`), which is nice because it's much simpler than when caching happens at multiple levels. I think it would be very nice to avoid caching happening inside the `metricsview` package – so the guarantee basically becomes that it always returns fresh data, which upstream callers can cache (we may need multi-level caching at some point, but I think we may be able to avoid it for now).

Looking at the resolvers, the `Cacheable` and `Key` functions are only called from here: https://github.com/rilldata/rill/blob/8e7bca1dc80ebeb7fb1f32df2312a02f0d4228e1/runtime/resolver.go#L126-L136.

What if we replaced these with a `CacheKey(ctx context.Context) (key []byte, ok bool)` function (where caching should be skipped if `!ok`)? Then the `metrics`, `metrics_sql` and `metrics_time_range` resolvers could call the `metrics_cache_key` resolver from their `CacheKey` implementation – if they pass that call through `opts.Runtime.Resolve`, then it in turn will also be cached.

Do you think that will work? The advantage of this approach is that there continues only to be caching happening in `runtime.Runtime.Resolve`, and resolvers are only being called from other resolvers (not from the `metricsview` package). This doesn't quite account for the legacy `runtime.Runtime.Query` implementation, but hopefully the above idea can also work for that in a similar way?

When it comes to the actual execution (but not caching) of the `metrics_cache_key` resolver, that in turn could actually be implemented inside the `metricsview` package (similar to how the logic for the `metrics` resolver is implemented there).

Lastly, to comment on this:
> Like as of now we are always querying watermark so it could be cached as well
You are thinking about this, right? https://github.com/rilldata/rill/blob/8e7bca1dc80ebeb7fb1f32df2312a02f0d4228e1/runtime/metricsview/executor_rewrite_time.go#L45

I agree that it would be nice to have this cached (it hasn't been urgent since the UI currently always passes fixed time ranges). In this case, we could also avoid multi-level caching by instead moving this to the resolver – e.g. you could have something like this in the `metrics` resolver (pseudocode):
```go
// Create an executor
executor := ...

// Call watermark resolver for the metrics view (can be cached in the resolver cache)
watermark := opts.Runtime.Resolve(...)

// Bind the query using the provided watermark (to avoid nested, uncached queries for the watermark)
qry := executor.BindWatermark(qry, watermark)

// Now execute the qry
res := executor.Execute(qry)
```

Reply:
Overall the `CacheKey` API seems good to me.
But I am somewhat unconvinced with how it is computed in resolvers.
1. Feels like too much indirection for `CacheKey`. It is hard to understand and trace the calls mentally. `metrics` and other resolvers need to call another resolver via the `runtime.Resolve` method to resolve cache key so that data can be cached. 
2. Some APIs on executor like `Query` can be called directly but other APIs like `CacheKey` need to go via the resolver path. So the decision to cache something is split across multiple places.
3. Overall the resolver construct is a generic construct but the results are stored in a `query_cache`. Though no resolver apart from the ones that operate on a metrics_view cache results as of now but the construct allows it. It may be easily fixable by calling the cache as `resolver_cache`.


May be for internal resolvers there should be a uncomplicated way to get and set things in query_cache to avoid too many indirections.

Reply:
> Feels like too much indirection for CacheKey

Agree, but I think it's intrinsic to the idea of a "cached cache key", not this implementation.

> Some APIs on executor like Query can be called directly but other APIs like CacheKey need to go via the resolver path. So the decision to cache something is split across multiple places.

You must have misunderstood `Executor.Query` – it never does any caching. So on the contrary, this ensures consistency because caching is *never* implemented in the `metricsview` package, and only implemented one place (in the `runtime/resolver.go` function (and the legacy `runtime/query.go`, but that's being deprecated)).

> It will be ideal to not let resolvers call other resolvers to prevent some cyclic calls.

I think resolvers calling other resolvers can actually be very powerful. Think of resolvers as memoized functions – it's completely fine for functions to call other functions, but of course, you should be careful if you call the same function recursively. 

> there should be a uncomplicated way to get and set things in query_cache to avoid too many indirections.

I'm conversely more worried about the potential for bugs if stuff is getting cached at many different levels in the call chain. Having caching only happen in `Resolve` IMO makes it easier to reason about when things are or are not cached.

Reply:
> Agree, but I think it's intrinsic to the idea of a "cached cache key", not this implementation.

Same indirection will be present if resolvers need to load watermark as well.

> You must have misunderstood Executor.Query – it never does any caching

No I didn't meant that. I meant `metrics` resolver and other such resolvers need to call `Executor.Query` directly but call `Executor.CacheKey` via `runtime.Resolve`.

> I think resolvers calling other resolvers can actually be very powerful. 

I mean there is no denying that it is powerful. My reservations are more around ease of usage and ease of understanding. Creating resolvers for internal usage feels like adding too much bloat. I think there has to be a tradeoff between "perfect" abstractions and ease of understanding/usage. May be we are now in _it is subjective_ territory.

> It's subjective, but I'm conversely more worried about the potential for bugs if stuff is getting cached at many different levels in the call chain.

I don't know if the current approach is less bug prone. It is easy to miss the fact calling the `Executor. CacheKey` in resolver needs to go via `runtime.Resolve` and not to be called directly which is not the case with other APIs on executor.

Reply:
I think this is all subjective, and we may need multiple levels of caching in the future, but as long as possible, I think we should aim to keep caching and invalidation in a single place (`Resolve`). If someone forgets to use it, the worst case is that something doesn't get cached until we notice.